Request event dispatcher

gavin-norman-sociomantic opened the pull request:
Based on #8. Blocked until merged.

Commit:
Add RequestOnConnBase.EventDispatcher.resumeFiber method

This can be useful for multi-fiber requests which are communicating with
each other via fiber resume codes of the request-on-conn.

gavin-norman-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -0,0 +1,690 @@
+                {
+                    this.dispatchQueuedSignals(conn);
+
+                    auto resume_code = conn.sendReceiveAndHandleEvents(
```
I think `sendReceiveAndHandleEvents` (also `receiveAndHandleEvents`, called below) can throw, in the case of an I/O error. How should these errors be dispatched?

gavin-norman-sociomantic replied:
Updated with handling for request-on-conn exceptions. (All waiting fibers are resumed with the exception.)

gavin-norman-sociomantic commented:
#8 merged. Unblocked.

gavin-norman-sociomantic commented:
Reviewer beware: the order of the commits displayed in github is not correct!

gavin-norman-sociomantic commented:
Correct commit order is:
1. Add RequestOnConnBase.EventDispatcher.resumeFiber method (https://github.com/sociomantic-tsunami/swarm/pull/19/commits/e6e4230119ab733ff01e134f80827f99d636778e)
2. Add RequestEventDispatcher helper struct (https://github.com/sociomantic-tsunami/swarm/pull/19/commits/ab53d7cd0308fb00e03850b1315045001471c62c)
3. Adapt example node's GetAll request to use worker/controller fibers (https://github.com/sociomantic-tsunami/swarm/pull/19/commits/c7c681d77b5eacddacb1e44d6a2abc89651b4066)
4. Remove commented out code (https://github.com/sociomantic-tsunami/swarm/pull/19/commits/214f5578da9cb0e2bea1546a04a710eedbc224aa)

david-eckardt-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -0,0 +1,767 @@
+{
+    /// The type of the message being awaited.
+    ubyte type;
+}
```
What is the reason for single-member structs like this -- why not a `ubyte message_type` field in `EventInUnion`?

gavin-norman-sociomantic replied:
True, could probably be simplified. (I think originally they had more fields.)

gavin-norman-sociomantic replied:
Ah actually there is a reason for this which became clear when I tried factoring it out: wrapping the union fields in structs allows for usage like this:
```D
ubyte awaited_message, awaited_signal;
RequestEventDispatcher dispatcher;
dispatcher.nextEvent(my_fiber, Message(awaited_message), Signal(awaited_signal));
```
With simple union fields, that would look like:
```D
ubyte awaited_message, awaited_signal;
EventIn message_event;
message_event.message = awaited_message;
EventIn signal_event;
signal_event.signal = awaited_signal;
RequestEventDispatcher dispatcher;
dispatcher.nextEvent(my_fiber, message_event, signal_event);
```

gavin-norman-sociomantic replied:
(The point is that the `signal` and `message` fields of the event union are both `ubyte`. With a plain union, there's no way to distinguish them other than directly setting them by name.)

david-eckardt-sociomantic replied:
Fair enough.

david-eckardt-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -0,0 +1,767 @@
+            default:
+                assert(false);
+        }
+
```
Interesting case, comparing two unions: Looks terribly redundant, but I don't see how it could be simplified.

david-eckardt-sociomantic commented on test/neo/client/request/internal/GetAll.d:
```diff
@@ -213,8 +181,178 @@ public struct GetAll
+                        count++;
+
+                        auto key = *conn.message_parser.
+                            getValue!(hash_t)(message.payload);
```
Why is the key not a hexadecimal string? :trollface:

david-eckardt-sociomantic commented on test/neo/client/request/internal/GetAll.d:
```diff
@@ -213,8 +181,178 @@ public struct GetAll
+            auto event = this.outer.request_event_dispatcher.nextEvent(
+                this.fiber, Signal(Signals.Stop));
+            assert(event.active == event.active.signal);
+            assert(event.signal.code == Signals.Stop);
```
The `in` contract of `event.signal()` contains ` assert(event.active == event.active.signal)`.

Commit:
Adapt example node's GetAll request to use worker/controller fibers

This provides an example of using the newly added RequestEventDispatcher in
a real request. A single control message -- Stop -- is supported currently
and the client's GetAll implementation is hacked to stop the iteration
after a fixed number of records have been received. A proper controller API
will be added in a future commit.

Commit:
Update to latest ocean patch release

(The subsequent commit requires the fix to
https://github.com/sociomantic-tsunami/ocean/issues/106 which is merged in
v3.1.4.)

gavin-norman-sociomantic commented:
Updated:
* Updated to ocean v3.1.4 and removed `toString` workarounds in event unions.
* Small doc enhancements.
* Removed duplicate assertion.

gavin-norman-sociomantic commented:
Updated:
* Added asserts to `RequestEventDispatcher.abort` to ensure that 1. a running fiber cannot be aborted, 2. a terminated fiber can have no events registered.

nemanja-boric-sociomantic commented on relnotes/resume-fiber.feature.md:
```diff
@@ -0,0 +1,8 @@
+* `swarm.neo.connection.RequestOnConnBase`
+
+  The nested class `EventDispatcher` now has the ability to suspend the outer
```
Strange how this talks about suspending the fiber, while this patch is adding `resumeFiber` method.

gavin-norman-sociomantic replied:
Ha good catch. Word mismatch.

gavin-norman-sociomantic replied:
Updated.

nemanja-boric-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -0,0 +1,776 @@
+
+*******************************************************************************/
+
+public alias SmartUnion!(EventInUnion) EventIn;
```
Where does the `In` coming from? Maybe some other name would be a better choice? I don't have any other suggestion, though.

gavin-norman-sociomantic replied:
Changed to `EventRegistration` and `EventNotification`. I think that's clearer.

nemanja-boric-sociomantic replied:
Much clearer, that's what I guessed this is, thanks!

nemanja-boric-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -0,0 +1,776 @@
+
+private union EventInUnion
+{
+    Message message;
```
Documentation.

nemanja-boric-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -0,0 +1,776 @@
+            if ( writer.fiber !is null && writer.fiber.waiting )
+                writer.fiber.resume(this.token, null, msg);
+
+            this.waiting_fibers.length = 0;
```
`enableStomping`?

nemanja-boric-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -0,0 +1,776 @@
+
+    private void unregister ( MessageFiber fiber, EventIn event )
+    {
+        this.waiting_fibers.length = Array.moveToEnd(this.waiting_fibers,
```
`enableStomping`?

gavin-norman-sociomantic commented:
Updated.

nemanja-boric-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -0,0 +1,777 @@
+    private EventNotification last_event;
+
+    /// Token passed to fiber suspend/resume calls.
+    private static MessageFiber.Token token;
```
This doesn't seem to be initialized anywhere, would this be a problem? (I presume no, as IIRC, we're using `MessageFiber` here solely for its ability to kill the fiber)

gavin-norman-sociomantic replied:
Yeah... I don't _think_ it's a problem to just use the default values. I'd like to split out the killing vs the token checking vs the message passing facilities of `MessageFiber`...

nemanja-boric-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -0,0 +1,777 @@
+        assert(signal <= 255);
+        ubyte signal_ubyte = cast(ubyte)signal;
+
+        foreach ( waiting_fiber; this.waiting_fibers )
```
One note here on the performance side (not sure if it's worth paying attention now, but it could be useful if it turns out that these loops are very tight when handling a lot of requests): Instead of a single list `this.waiting_fibers` we could have `this.fibers_waiting_for_message` and `this.fibers_waiting_for_signal` and then even these could actually be arrays of arrays, where the first dimension is the message/signal code, so you can access the first waiting fiber in constant time.

gavin-norman-sociomantic replied:
I considered that, but remember that this is an event dispatcher _for a single request_. So we'd only need to care about this kind of optimisation if we get to the stage of having requests with dozens of fibers. If that ever happens, I think we'll have other problems to worry about too ;)

nemanja-boric-sociomantic replied:
Ah, yes, very good point - this is an event dispatcher for a single request. All good :+1:

nemanja-boric-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -0,0 +1,777 @@
+            }
+        }
+
+        log.warn("Unhandled message of type {}", message_type);
```
Last nitpick: to me it feels that this is more useful for the debugging, than it is for the live servers, and it feels in the case of the misbehaved client this could come quite costly, as there could be a big pressure on the logging system, and it could lead to the denial of service attack. I suggest either we remove this, or move it to the `info` level, so it will not be logged by default on live configurations.

gavin-norman-sociomantic replied:
Hm, it's a tricky point. I deliberately made this a relatively high log level as it really should be noticed if this ever happens. (I'd envisaged using this in combination with the log level stats writer in ocean.)

I see what you mean, though. If the time between the buggy request starting, this error occurring, and the client firing the request again was very small, it could lead to a big performance hit affecting other requests.

nemanja-boric-sociomantic replied:
If this is the concern, we could (in separate PR) setup stats so it logs count of how many times the connection has been shut down with the protocol error.

gavin-norman-sociomantic replied:
You're right. That would be a better solution.

gavin-norman-sociomantic replied:
Upated.

nemanja-boric-sociomantic reviewed (approved):
LGTM, thanks!

Commit:
Add RequestEventDispatcher helper struct

Multi-fiber requests require an event dispatcher to multiplex events from a
single request-on-conn over multiple fibers.

Fixes #14.

gavin-norman-sociomantic commented:
(Also updated with a non-default fiber token, just for usage correctness.)
